### ISSUE TITLE
Make RootContext immutable

### ISSUE THREAD

**Issue by timbussmann** [Particular/NServiceBus#4580]
As we have seen inhttps://github.com/Particular/NServiceBus/issues/4573, there is the possibility of some code accidentally (or intentionally) modifying/adding values to the root context. As the `RootContext` class is used as a base class for every incoming or outgoing pipeline context, this might have a huge impact on the whole endpoint.

For safety, we should prevent this risk by making the `RootContext` immutable in some way. e.g.
* detect and throw on changes
* provide no api to change/add values
* do not share a single `RootContext`

**Comment by timbussmann** [comment]
good point brought up by @bording: we should probably also verify/test that we do not accidentally set/leak into parent context when creating new behavior contexts in the different pipeline stages.

**Comment by mikeminutillo** [comment]
Is this limited to RootContext? Would it make sense to allow any context to be frozen and cause any modification to it to throw? In #4573, it was possible for the parent context (a handler invoke) to have things incorrectly put into it. 

As soon as something becomes a Parent context, it doesn't make sense to add anything to it. Any downstream dependencies are usually reading objects from the parent context and setting some state on those if found. Should Parent Context hide behind a readonly interface?

**Comment by timbussmann** [comment]
this is not strictly limited to rootcontext but the other contexts cannot be completely frozen, e.g.:

StageContextA is parrent of StageContextB. During stage B, context A must not be modified, but in the behaviors in stage A, context A must be modifiable all the times. There is little/no risk in leaking settings/data between operations that's why it's probably not worth in going beyond testing the correct behavior.

**Comment by timbussmann** [comment]
Marking this a breaking change as it seems it would be tricky to implement this in a minor and I don't really see much value in that. It would be best to incorporate the concept of a read-only root context into the public API.
